File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit_objc/FIAObjectTranslator.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "./include/in_app_purchase_storekit_objc/FIAObjectTranslator.h"

#pragma mark - SKProduct Coders

@implementation FIAObjectTranslator

+ (NSDictionary *)getMapFromSKProduct:(SKProduct *)product {
  if (!product) {
    return nil;
  }
  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"localizedDescription" : product.localizedDescription ?: [NSNull null],
    @"localizedTitle" : product.localizedTitle ?: [NSNull null],
    @"productIdentifier" : product.productIdentifier ?: [NSNull null],
    @"price" : product.price.description ?: [NSNull null]

  }];
  // TODO(cyanglaz): NSLocale is a complex object, want to see the actual need of getting this
  // expanded to a map. Matching android to only get the currencySymbol for now.
  // https://github.com/flutter/flutter/issues/26610
  [map setObject:[FIAObjectTranslator getMapFromNSLocale:product.priceLocale] ?: [NSNull null]
          forKey:@"priceLocale"];
  [map setObject:[FIAObjectTranslator
                     getMapFromSKProductSubscriptionPeriod:product.subscriptionPeriod]
                     ?: [NSNull null]
          forKey:@"subscriptionPeriod"];
  [map setObject:[FIAObjectTranslator getMapFromSKProductDiscount:product.introductoryPrice]
                     ?: [NSNull null]
          forKey:@"introductoryPrice"];
  if (@available(iOS 12.2, *)) {
    [map setObject:[FIAObjectTranslator getMapArrayFromSKProductDiscounts:product.discounts]
            forKey:@"discounts"];
  }
  [map setObject:product.subscriptionGroupIdentifier ?: [NSNull null]
          forKey:@"subscriptionGroupIdentifier"];
  return map;
}

+ (NSDictionary *)getMapFromSKProductSubscriptionPeriod:(SKProductSubscriptionPeriod *)period {
  if (!period) {
    return nil;
  }
  return @{@"numberOfUnits" : @(period.numberOfUnits), @"unit" : @(period.unit)};
}

+ (nonnull NSArray *)getMapArrayFromSKProductDiscounts:
    (nonnull NSArray<SKProductDiscount *> *)productDiscounts {
  NSMutableArray *discountsMapArray = [[NSMutableArray alloc] init];

  for (SKProductDiscount *productDiscount in productDiscounts) {
    [discountsMapArray addObject:[FIAObjectTranslator getMapFromSKProductDiscount:productDiscount]];
  }

  return discountsMapArray;
}

+ (NSDictionary *)getMapFromSKProductDiscount:(SKProductDiscount *)discount {
  if (!discount) {
    return nil;
  }
  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"price" : discount.price.description ?: [NSNull null],
    @"numberOfPeriods" : @(discount.numberOfPeriods),
    @"subscriptionPeriod" :
            [FIAObjectTranslator getMapFromSKProductSubscriptionPeriod:discount.subscriptionPeriod]
        ?: [NSNull null],
    @"paymentMode" : @(discount.paymentMode),
  }];
  if (@available(iOS 12.2, *)) {
    [map setObject:discount.identifier ?: [NSNull null] forKey:@"identifier"];
    [map setObject:@(discount.type) forKey:@"type"];
  }

  // TODO(cyanglaz): NSLocale is a complex object, want to see the actual need of getting this
  // expanded to a map. Matching android to only get the currencySymbol for now.
  // https://github.com/flutter/flutter/issues/26610
  [map setObject:[FIAObjectTranslator getMapFromNSLocale:discount.priceLocale] ?: [NSNull null]
          forKey:@"priceLocale"];
  return map;
}

+ (NSDictionary *)getMapFromSKProductsResponse:(SKProductsResponse *)productResponse {
  if (!productResponse) {
    return nil;
  }
  NSMutableArray *productsMapArray = [NSMutableArray new];
  for (SKProduct *product in productResponse.products) {
    [productsMapArray addObject:[FIAObjectTranslator getMapFromSKProduct:product]];
  }
  return @{
    @"products" : productsMapArray,
    @"invalidProductIdentifiers" : productResponse.invalidProductIdentifiers ?: @[]
  };
}

+ (NSDictionary *)getMapFromSKPayment:(SKPayment *)payment {
  if (!payment) {
    return nil;
  }
  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"productIdentifier" : payment.productIdentifier ?: [NSNull null],
    @"requestData" : payment.requestData ? [[NSString alloc] initWithData:payment.requestData
                                                                 encoding:NSUTF8StringEncoding]
                                         : [NSNull null],
    @"quantity" : @(payment.quantity),
    @"applicationUsername" : payment.applicationUsername ?: [NSNull null]
  }];
  [map setObject:@(payment.simulatesAskToBuyInSandbox) forKey:@"simulatesAskToBuyInSandbox"];
  return map;
}

+ (NSDictionary *)getMapFromNSLocale:(NSLocale *)locale {
  if (!locale) {
    return nil;
  }
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  [map setObject:locale.currencySymbol ?: [NSNull null] forKey:@"currencySymbol"];
  [map setObject:locale.currencyCode ?: [NSNull null] forKey:@"currencyCode"];
  [map setObject:locale.countryCode ?: [NSNull null] forKey:@"countryCode"];
  return map;
}

+ (SKMutablePayment *)getSKMutablePaymentFromMap:(NSDictionary *)map {
  if (!map) {
    return nil;
  }
  SKMutablePayment *payment = [[SKMutablePayment alloc] init];
  payment.productIdentifier = map[@"productIdentifier"];
  NSString *utf8String = map[@"requestData"];
  payment.requestData = [utf8String dataUsingEncoding:NSUTF8StringEncoding];
  payment.quantity = [map[@"quantity"] integerValue];
  payment.applicationUsername = map[@"applicationUsername"];
  payment.simulatesAskToBuyInSandbox = [map[@"simulatesAskToBuyInSandbox"] boolValue];
  return payment;
}

+ (NSDictionary *)getMapFromSKPaymentTransaction:(SKPaymentTransaction *)transaction {
  if (!transaction) {
    return nil;
  }
  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"error" : [FIAObjectTranslator getMapFromNSError:transaction.error] ?: [NSNull null],
    @"payment" : transaction.payment ? [FIAObjectTranslator getMapFromSKPayment:transaction.payment]
                                     : [NSNull null],
    @"originalTransaction" : transaction.originalTransaction
        ? [FIAObjectTranslator getMapFromSKPaymentTransaction:transaction.originalTransaction]
        : [NSNull null],
    @"transactionTimeStamp" : transaction.transactionDate
        ? @(transaction.transactionDate.timeIntervalSince1970)
        : [NSNull null],
    @"transactionIdentifier" : transaction.transactionIdentifier ?: [NSNull null],
    @"transactionState" : @(transaction.transactionState)
  }];

  return map;
}

+ (NSDictionary *)getMapFromNSError:(NSError *)error {
  if (!error) {
    return nil;
  }

  return @{
    @"code" : @(error.code),
    @"domain" : error.domain ?: @"",
    @"userInfo" : [FIAObjectTranslator encodeNSErrorUserInfo:error.userInfo]
  };
}

+ (id)encodeNSErrorUserInfo:(id)value {
  if ([value isKindOfClass:[NSError class]]) {
    return [FIAObjectTranslator getMapFromNSError:value];
  } else if ([value isKindOfClass:[NSURL class]]) {
    return [value absoluteString];
  } else if ([value isKindOfClass:[NSNumber class]]) {
    return value;
  } else if ([value isKindOfClass:[NSString class]]) {
    return value;
  } else if ([value isKindOfClass:[NSArray class]]) {
    NSMutableArray *errors = [[NSMutableArray alloc] init];
    for (id error in value) {
      [errors addObject:[FIAObjectTranslator encodeNSErrorUserInfo:error]];
    }
    return errors;
  } else if ([value isKindOfClass:[NSDictionary class]]) {
    NSMutableDictionary *errors = [[NSMutableDictionary alloc] init];
    for (id key in value) {
      errors[key] = [FIAObjectTranslator encodeNSErrorUserInfo:value[key]];
    }
    return errors;
  } else {
    return [NSString
        stringWithFormat:
            @"Unable to encode native userInfo object of type %@ to map. Please submit an issue at "
            @"https://github.com/flutter/flutter/issues/new with the title "
            @"\"[in_app_purchase_storekit] "
            @"Unable to encode userInfo of type %@\" and add reproduction steps and the error "
            @"details in "
            @"the description field.",
            [value class], [value class]];
  }
}

+ (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront {
  if (!storefront) {
    return nil;
  }

  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"countryCode" : storefront.countryCode,
    @"identifier" : storefront.identifier
  }];

  return map;
}

+ (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
                 andSKPaymentTransaction:(SKPaymentTransaction *)transaction {
  if (!storefront || !transaction) {
    return nil;
  }

  NSMutableDictionary *map = [[NSMutableDictionary alloc] initWithDictionary:@{
    @"storefront" : [FIAObjectTranslator getMapFromSKStorefront:storefront],
    @"transaction" : [FIAObjectTranslator getMapFromSKPaymentTransaction:transaction]
  }];

  return map;
}

+ (SKPaymentDiscount *)getSKPaymentDiscountFromMap:(NSDictionary *)map
                                         withError:(NSString **)error {
  if (!map || map.count <= 0) {
    return nil;
  }

  NSString *identifier = map[@"identifier"];
  NSString *keyIdentifier = map[@"keyIdentifier"];
  NSString *nonce = map[@"nonce"];
  NSString *signature = map[@"signature"];
  NSNumber *timestamp = map[@"timestamp"];

  if (!identifier || ![identifier isKindOfClass:NSString.class] ||
      [identifier isEqualToString:@""]) {
    if (error) {
      *error = @"When specifying a payment discount the 'identifier' field is mandatory.";
    }
    return nil;
  }

  if (!keyIdentifier || ![keyIdentifier isKindOfClass:NSString.class] ||
      [keyIdentifier isEqualToString:@""]) {
    if (error) {
      *error = @"When specifying a payment discount the 'keyIdentifier' field is mandatory.";
    }
    return nil;
  }

  if (!nonce || ![nonce isKindOfClass:NSString.class] || [nonce isEqualToString:@""]) {
    if (error) {
      *error = @"When specifying a payment discount the 'nonce' field is mandatory.";
    }
    return nil;
  }

  if (!signature || ![signature isKindOfClass:NSString.class] || [signature isEqualToString:@""]) {
    if (error) {
      *error = @"When specifying a payment discount the 'signature' field is mandatory.";
    }
    return nil;
  }

  if (!timestamp || ![timestamp isKindOfClass:NSNumber.class] || [timestamp longLongValue] <= 0) {
    if (error) {
      *error = @"When specifying a payment discount the 'timestamp' field is mandatory.";
    }
    return nil;
  }

  SKPaymentDiscount *discount =
      [[SKPaymentDiscount alloc] initWithIdentifier:identifier
                                      keyIdentifier:keyIdentifier
                                              nonce:[[NSUUID alloc] initWithUUIDString:nonce]
                                          signature:signature
                                          timestamp:timestamp];

  return discount;
}

+ (nullable FIASKPaymentTransactionMessage *)convertTransactionToPigeon:
    (nullable SKPaymentTransaction *)transaction API_AVAILABLE(ios(12.2)) {
  if (!transaction) {
    return nil;
  }
  return [FIASKPaymentTransactionMessage
            makeWithPayment:[self convertPaymentToPigeon:transaction.payment]
           transactionState:[self convertTransactionStateToPigeon:transaction.transactionState]
        originalTransaction:transaction.originalTransaction
                                ? [self convertTransactionToPigeon:transaction.originalTransaction]
                                : nil
       transactionTimeStamp:[NSNumber numberWithDouble:[transaction.transactionDate
                                                               timeIntervalSince1970]]
      transactionIdentifier:transaction.transactionIdentifier
                      error:[self convertSKErrorToPigeon:transaction.error]];
}

+ (nullable FIASKErrorMessage *)convertSKErrorToPigeon:(nullable NSError *)error {
  if (!error) {
    return nil;
  }

  NSMutableDictionary *userInfo = [NSMutableDictionary new];
  for (NSErrorUserInfoKey key in error.userInfo) {
    id value = error.userInfo[key];
    userInfo[key] = [FIAObjectTranslator encodeNSErrorUserInfo:value];
  }

  return [FIASKErrorMessage makeWithCode:error.code domain:error.domain userInfo:userInfo];
}

+ (FIASKPaymentTransactionStateMessage)convertTransactionStateToPigeon:
    (SKPaymentTransactionState)state {
  switch (state) {
    case SKPaymentTransactionStatePurchasing:
      return FIASKPaymentTransactionStateMessagePurchasing;
    case SKPaymentTransactionStatePurchased:
      return FIASKPaymentTransactionStateMessagePurchased;
    case SKPaymentTransactionStateFailed:
      return FIASKPaymentTransactionStateMessageFailed;
    case SKPaymentTransactionStateRestored:
      return FIASKPaymentTransactionStateMessageRestored;
    case SKPaymentTransactionStateDeferred:
      return FIASKPaymentTransactionStateMessageDeferred;
  }
}

+ (nullable FIASKPaymentMessage *)convertPaymentToPigeon:(nullable SKPayment *)payment
    API_AVAILABLE(ios(12.2)) {
  if (!payment) {
    return nil;
  }
  return [FIASKPaymentMessage
       makeWithProductIdentifier:payment.productIdentifier
             applicationUsername:payment.applicationUsername
                     requestData:[[NSString alloc] initWithData:payment.requestData
                                                       encoding:NSUTF8StringEncoding]
                        quantity:payment.quantity
      simulatesAskToBuyInSandbox:payment.simulatesAskToBuyInSandbox
                 paymentDiscount:[self convertPaymentDiscountToPigeon:payment.paymentDiscount]];
}

+ (nullable FIASKPaymentDiscountMessage *)convertPaymentDiscountToPigeon:
    (nullable SKPaymentDiscount *)discount API_AVAILABLE(ios(12.2)) {
  if (!discount) {
    return nil;
  }
  return [FIASKPaymentDiscountMessage makeWithIdentifier:discount.identifier
                                           keyIdentifier:discount.keyIdentifier
                                                   nonce:[discount.nonce UUIDString]
                                               signature:discount.signature
                                               timestamp:[discount.timestamp intValue]];
}

+ (nullable FIASKStorefrontMessage *)convertStorefrontToPigeon:(nullable SKStorefront *)storefront
    API_AVAILABLE(ios(13.0)) {
  if (!storefront) {
    return nil;
  }
  return [FIASKStorefrontMessage makeWithCountryCode:storefront.countryCode
                                          identifier:storefront.identifier];
}

+ (nullable FIASKProductSubscriptionPeriodMessage *)convertSKProductSubscriptionPeriodToPigeon:
    (nullable SKProductSubscriptionPeriod *)period API_AVAILABLE(ios(12.2)) {
  if (!period) {
    return nil;
  }

  FIASKSubscriptionPeriodUnitMessage unit;
  switch (period.unit) {
    case SKProductPeriodUnitDay:
      unit = FIASKSubscriptionPeriodUnitMessageDay;
      break;
    case SKProductPeriodUnitWeek:
      unit = FIASKSubscriptionPeriodUnitMessageWeek;
      break;
    case SKProductPeriodUnitMonth:
      unit = FIASKSubscriptionPeriodUnitMessageMonth;
      break;
    case SKProductPeriodUnitYear:
      unit = FIASKSubscriptionPeriodUnitMessageYear;
      break;
  }

  return [FIASKProductSubscriptionPeriodMessage makeWithNumberOfUnits:period.numberOfUnits
                                                                 unit:unit];
}

+ (nullable FIASKProductDiscountMessage *)convertProductDiscountToPigeon:
    (nullable SKProductDiscount *)productDiscount API_AVAILABLE(ios(12.2)) {
  if (!productDiscount) {
    return nil;
  }

  FIASKProductDiscountPaymentModeMessage paymentMode;
  switch (productDiscount.paymentMode) {
    case SKProductDiscountPaymentModeFreeTrial:
      paymentMode = FIASKProductDiscountPaymentModeMessageFreeTrial;
      break;
    case SKProductDiscountPaymentModePayAsYouGo:
      paymentMode = FIASKProductDiscountPaymentModeMessagePayAsYouGo;
      break;
    case SKProductDiscountPaymentModePayUpFront:
      paymentMode = FIASKProductDiscountPaymentModeMessagePayUpFront;
      break;
  }

  FIASKProductDiscountTypeMessage type;
  switch (productDiscount.type) {
    case SKProductDiscountTypeIntroductory:
      type = FIASKProductDiscountTypeMessageIntroductory;
      break;
    case SKProductDiscountTypeSubscription:
      type = FIASKProductDiscountTypeMessageSubscription;
      break;
  }

  return [FIASKProductDiscountMessage
           makeWithPrice:productDiscount.price.description
             priceLocale:[self convertNSLocaleToPigeon:productDiscount.priceLocale]
         numberOfPeriods:productDiscount.numberOfPeriods
             paymentMode:paymentMode
      subscriptionPeriod:[self convertSKProductSubscriptionPeriodToPigeon:productDiscount
                                                                              .subscriptionPeriod]
              identifier:productDiscount.identifier
                    type:type];
}

+ (nullable FIASKPriceLocaleMessage *)convertNSLocaleToPigeon:(nullable NSLocale *)locale
    API_AVAILABLE(ios(12.2)) {
  if (!locale) {
    return nil;
  }
  return [FIASKPriceLocaleMessage makeWithCurrencySymbol:locale.currencySymbol
                                            currencyCode:locale.currencyCode
                                             countryCode:locale.countryCode];
}

+ (nullable FIASKProductMessage *)convertProductToPigeon:(nullable SKProduct *)product
    API_AVAILABLE(ios(12.2)) {
  if (!product) {
    return nil;
  }

  NSArray<SKProductDiscount *> *skProductDiscounts = product.discounts;
  NSMutableArray<FIASKProductDiscountMessage *> *pigeonProductDiscounts =
      [NSMutableArray arrayWithCapacity:skProductDiscounts.count];

  for (SKProductDiscount *productDiscount in skProductDiscounts) {
    [pigeonProductDiscounts addObject:[self convertProductDiscountToPigeon:productDiscount]];
  };

  return [FIASKProductMessage
        makeWithProductIdentifier:product.productIdentifier
                   localizedTitle:product.localizedTitle
             localizedDescription:product.localizedDescription
                      priceLocale:[self convertNSLocaleToPigeon:product.priceLocale]
      subscriptionGroupIdentifier:product.subscriptionGroupIdentifier
                            price:product.price.description
               subscriptionPeriod:
                   [self convertSKProductSubscriptionPeriodToPigeon:product.subscriptionPeriod]
                introductoryPrice:[self convertProductDiscountToPigeon:product.introductoryPrice]
                        discounts:pigeonProductDiscounts];
}

+ (nullable FIASKProductsResponseMessage *)convertProductsResponseToPigeon:
    (nullable SKProductsResponse *)productsResponse API_AVAILABLE(ios(12.2)) {
  if (!productsResponse) {
    return nil;
  }
  NSArray<SKProduct *> *skProducts = productsResponse.products;
  NSMutableArray<FIASKProductMessage *> *pigeonProducts =
      [NSMutableArray arrayWithCapacity:skProducts.count];

  for (SKProduct *product in skProducts) {
    [pigeonProducts addObject:[self convertProductToPigeon:product]];
  };

  return [FIASKProductsResponseMessage
               makeWithProducts:pigeonProducts
      invalidProductIdentifiers:productsResponse.invalidProductIdentifiers ?: @[]];
}

@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/FIAObjectTranslator.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import StoreKit

#if canImport(in_app_purchase_storekit_objc)
  import in_app_purchase_storekit_objc
#endif

public class FIAObjectTranslator: NSObject {
  // MARK: - SKProduct Coders

  public static func getMapFrom(_ product: SKProduct) -> [String: Any] {
    return [
      "discounts": getMapArrayFrom(product.discounts),
      "introductoryPrice": product.introductoryPrice.map { getMapFrom($0) } as Any,
      "localizedDescription": product.localizedDescription,
      "localizedTitle": product.localizedTitle,
      "productIdentifier": product.productIdentifier,
      "price": product.price.description,
      "subscriptionGroupIdentifier": product.subscriptionGroupIdentifier as Any,
      "subscriptionPeriod": product.subscriptionPeriod.map { getMapFrom($0) } as Any,
      "priceLocale": getMapFrom(product.priceLocale),
    ]
  }

  public static func getMapFrom(_ period: SKProductSubscriptionPeriod) -> [String: Any] {
    return ["numberOfUnits": period.numberOfUnits, "unit": period.unit.rawValue]
  }

  static func getMapArrayFrom(_ productDiscounts: [SKProductDiscount]) -> [Any] {
    return productDiscounts.map { getMapFrom($0) }
  }

  public static func getMapFrom(_ discount: SKProductDiscount) -> [String: Any] {
    return [
      "identifier": discount.identifier as Any,
      "numberOfPeriods": discount.numberOfPeriods,
      "paymentMode": discount.paymentMode.rawValue,
      "price": discount.price.description,
      "subscriptionPeriod": getMapFrom(discount.subscriptionPeriod),
      "type": discount.type.rawValue,
      "priceLocale": getMapFrom(discount.priceLocale),
    ]
  }

  public static func getMapFrom(_ productResponse: SKProductsResponse) -> [String: Any] {
    let productsMapArray = productResponse.products.map { getMapFrom($0) }
    return [
      "products": productsMapArray,
      "invalidProductIdentifiers": productResponse.invalidProductIdentifiers,
    ]
  }

  public static func getMapFrom(_ payment: SKPayment) -> [String: Any] {
    return [
      "applicationUsername": payment.applicationUsername as Any,
      "productIdentifier": payment.productIdentifier as Any,
      "quantity": payment.quantity,
      "requestData": payment.requestData.flatMap { String(data: $0, encoding: .utf8) } as Any,
      "simulatesAskToBuyInSandbox": payment.simulatesAskToBuyInSandbox,
    ]
  }

  // This intentionally only exposes fields that there has been a demonstrated
  // need for; see discussion in https://github.com/flutter/plugins/pull/3897.
  public static func getMapFrom(_ locale: Locale) -> [String: Any] {
    let nsLocale = locale as NSLocale
    return [
      "currencySymbol": nsLocale.object(forKey: .currencySymbol) as Any,
      "currencyCode": nsLocale.object(forKey: .currencyCode) as Any,
      "countryCode": nsLocale.object(forKey: .countryCode) as Any,
    ]
  }

  public static func getSKMutablePayment(fromMap map: [String: Any]) -> SKMutablePayment {
    let payment = SKMutablePayment()
    payment.productIdentifier = map["productIdentifier"] as? String ?? ""
    if let utf8String = map["requestData"] as? String {
      payment.requestData = utf8String.data(using: .utf8)
    }
    payment.quantity = (map["quantity"] as? NSNumber)?.intValue ?? 0
    payment.applicationUsername = map["applicationUsername"] as? String
    payment.simulatesAskToBuyInSandbox =
      (map["simulatesAskToBuyInSandbox"] as? NSNumber)?.boolValue ?? false
    return payment
  }

  public static func getMapFrom(_ transaction: SKPaymentTransaction) -> [String: Any] {
    return [
      "error": transaction.error.map { getMapFrom($0 as NSError) } as Any,
      "payment": (transaction.value(forKey: "payment") as? SKPayment).map { getMapFrom($0) } as Any,
      "originalTransaction": transaction.original.map { getMapFrom($0) } as Any,
      "transactionTimeStamp": transaction.transactionDate?.timeIntervalSince1970 as Any,
      "transactionIdentifier": transaction.transactionIdentifier as Any,
      "transactionState": transaction.transactionState.rawValue,
    ]
  }

  public static func getMapFrom(_ error: NSError) -> [String: Any] {
    return [
      "code": error.code,
      "domain": error.domain,
      "userInfo": encodeNSErrorUserInfo(error.userInfo),
    ]
  }

  static func encodeNSErrorUserInfo(_ value: Any) -> Any {
    switch value {
    case let error as NSError:
      return getMapFrom(error)
    case let url as URL:
      return url.absoluteString
    case is NSNumber, is String:
      return value
    case let array as [Any]:
      return array.map { encodeNSErrorUserInfo($0) }
    case let dictionary as [AnyHashable: Any]:
      var errors: [AnyHashable: Any] = [:]
      for (key, dictValue) in dictionary {
        errors[key] = encodeNSErrorUserInfo(dictValue)
      }
      return errors
    default:
      return
        "Unable to encode native userInfo object of type \(type(of: value)) to map. Please submit an issue at "
        + "https://github.com/flutter/flutter/issues/new with the title "
        + "\"[in_app_purchase_storekit] "
        + "Unable to encode userInfo of type \(type(of: value))\" and add reproduction steps and the error "
        + "details in "
        + "the description field."
    }
  }

  public static func getMapFrom(_ storefront: SKStorefront) -> [String: Any] {
    return ["countryCode": storefront.countryCode, "identifier": storefront.identifier]
  }

  public static func getMapFrom(
    _ storefront: SKStorefront, andSKPaymentTransaction transaction: SKPaymentTransaction
  ) -> [String: Any] {
    return [
      "storefront": getMapFrom(storefront),
      "transaction": getMapFrom(transaction),
    ]
  }

  public static func getSKPaymentDiscount(
    fromMap map: [String: Any]?, withError error: inout NSString?
  )
    -> SKPaymentDiscount?
  {
    guard let map = map, !map.isEmpty else {
      return nil
    }

    let identifier = map["identifier"] as? String
    let keyIdentifier = map["keyIdentifier"] as? String
    let nonce = map["nonce"] as? String
    let signature = map["signature"] as? String
    let timestamp = map["timestamp"] as? NSNumber

    guard let identifier = identifier, !identifier.isEmpty else {
      error = "When specifying a payment discount the 'identifier' field is mandatory."
      return nil
    }

    guard let keyIdentifier = keyIdentifier, !keyIdentifier.isEmpty else {
      error = "When specifying a payment discount the 'keyIdentifier' field is mandatory."
      return nil
    }

    guard let nonce = nonce, !nonce.isEmpty else {
      error = "When specifying a payment discount the 'nonce' field is mandatory."
      return nil
    }

    guard let signature = signature, !signature.isEmpty else {
      error = "When specifying a payment discount the 'signature' field is mandatory."
      return nil
    }

    guard let timestamp = timestamp, timestamp.int64Value > 0 else {
      error = "When specifying a payment discount the 'timestamp' field is mandatory."
      return nil
    }

    guard let nonceUUID = UUID(uuidString: nonce) else {
      error = "When specifying a payment discount the 'nonce' field is mandatory."
      return nil
    }

    return SKPaymentDiscount(
      identifier: identifier, keyIdentifier: keyIdentifier, nonce: nonceUUID, signature: signature,
      timestamp: timestamp)
  }

  // MARK: - Pigeon message translators

  public static func convertTransaction(toPigeon transaction: SKPaymentTransaction?)
    -> FIASKPaymentTransactionMessage?
  {
    guard let transaction = transaction else {
      return nil
    }
    let paymentMessage =
      convertPayment(toPigeon: transaction.value(forKey: "payment") as? SKPayment)
      ?? FIASKPaymentMessage.make(
        withProductIdentifier: "", applicationUsername: nil, requestData: nil, quantity: 0,
        simulatesAskToBuyInSandbox: false, paymentDiscount: nil)
    return FIASKPaymentTransactionMessage.make(
      withPayment: paymentMessage,
      transactionState: convertTransactionStateToPigeon(transaction.transactionState),
      originalTransaction: transaction.original.flatMap {
        convertTransaction(toPigeon: $0)
      },
      transactionTimeStamp: NSNumber(
        value: transaction.transactionDate?.timeIntervalSince1970 ?? 0),
      transactionIdentifier: transaction.transactionIdentifier,
      error: convertSKError(toPigeon: transaction.error as NSError?))
  }

  public static func convertSKError(toPigeon error: NSError?) -> FIASKErrorMessage? {
    guard let error = error else {
      return nil
    }

    var userInfo: [String: Any] = [:]
    for (key, value) in error.userInfo {
      userInfo[key as? String ?? "\(key)"] = encodeNSErrorUserInfo(value)
    }

    return FIASKErrorMessage.make(withCode: error.code, domain: error.domain, userInfo: userInfo)
  }

  static func convertTransactionStateToPigeon(_ state: SKPaymentTransactionState)
    -> FIASKPaymentTransactionStateMessage
  {
    switch state {
    case .purchasing:
      return .purchasing
    case .purchased:
      return .purchased
    case .failed:
      return .failed
    case .restored:
      return .restored
    case .deferred:
      return .deferred
    @unknown default:
      return .purchasing
    }
  }

  public static func convertPayment(toPigeon payment: SKPayment?) -> FIASKPaymentMessage? {
    guard let payment = payment else {
      return nil
    }
    return FIASKPaymentMessage.make(
      withProductIdentifier: payment.productIdentifier,
      applicationUsername: payment.applicationUsername,
      requestData: payment.requestData.flatMap { String(data: $0, encoding: .utf8) },
      quantity: payment.quantity,
      simulatesAskToBuyInSandbox: payment.simulatesAskToBuyInSandbox,
      paymentDiscount: convertPaymentDiscount(toPigeon: payment.paymentDiscount))
  }

  public static func convertPaymentDiscount(toPigeon discount: SKPaymentDiscount?)
    -> FIASKPaymentDiscountMessage?
  {
    guard let discount = discount else {
      return nil
    }
    return FIASKPaymentDiscountMessage.make(
      withIdentifier: discount.identifier, keyIdentifier: discount.keyIdentifier,
      nonce: discount.nonce.uuidString, signature: discount.signature,
      timestamp: discount.timestamp.intValue)
  }

  public static func convertStorefront(toPigeon storefront: SKStorefront?)
    -> FIASKStorefrontMessage?
  {
    guard let storefront = storefront else {
      return nil
    }
    return FIASKStorefrontMessage.make(
      withCountryCode: storefront.countryCode, identifier: storefront.identifier)
  }

  public static func convertSKProductSubscriptionPeriod(
    toPigeon period: SKProductSubscriptionPeriod?
  ) -> FIASKProductSubscriptionPeriodMessage? {
    guard let period = period else {
      return nil
    }

    let unit: FIASKSubscriptionPeriodUnitMessage
    switch period.unit {
    case .day:
      unit = .day
    case .week:
      unit = .week
    case .month:
      unit = .month
    case .year:
      unit = .year
    @unknown default:
      unit = .day
    }

    return FIASKProductSubscriptionPeriodMessage.make(
      withNumberOfUnits: period.numberOfUnits, unit: unit)
  }

  public static func convertProductDiscount(toPigeon productDiscount: SKProductDiscount?)
    -> FIASKProductDiscountMessage?
  {
    guard let productDiscount = productDiscount else {
      return nil
    }

    let paymentMode: FIASKProductDiscountPaymentModeMessage
    switch productDiscount.paymentMode {
    case .freeTrial:
      paymentMode = .freeTrial
    case .payAsYouGo:
      paymentMode = .payAsYouGo
    case .payUpFront:
      paymentMode = .payUpFront
    @unknown default:
      paymentMode = .payAsYouGo
    }

    let type: FIASKProductDiscountTypeMessage
    switch productDiscount.type {
    case .introductory:
      type = .introductory
    case .subscription:
      type = .subscription
    @unknown default:
      type = .introductory
    }

    return FIASKProductDiscountMessage.make(
      withPrice: productDiscount.price.description,
      priceLocale: convertNSLocale(toPigeon: productDiscount.priceLocale)!,
      numberOfPeriods: productDiscount.numberOfPeriods,
      paymentMode: paymentMode,
      subscriptionPeriod: convertSKProductSubscriptionPeriod(
        toPigeon: productDiscount.subscriptionPeriod)!,
      identifier: productDiscount.identifier,
      type: type)
  }

  public static func convertNSLocale(toPigeon locale: Locale?) -> FIASKPriceLocaleMessage? {
    guard let locale = locale else {
      return nil
    }
    let nsLocale = locale as NSLocale
    return FIASKPriceLocaleMessage.make(
      withCurrencySymbol: nsLocale.object(forKey: .currencySymbol) as? String ?? "",
      currencyCode: nsLocale.object(forKey: .currencyCode) as? String ?? "",
      countryCode: nsLocale.object(forKey: .countryCode) as? String ?? "")
  }

  public static func convertProduct(toPigeon product: SKProduct?) -> FIASKProductMessage? {
    guard let product = product else {
      return nil
    }

    let pigeonProductDiscounts = product.discounts.map { convertProductDiscount(toPigeon: $0)! }

    return FIASKProductMessage.make(
      withProductIdentifier: product.productIdentifier,
      localizedTitle: product.localizedTitle,
      localizedDescription: product.localizedDescription,
      priceLocale: convertNSLocale(toPigeon: product.priceLocale)!,
      subscriptionGroupIdentifier: product.subscriptionGroupIdentifier,
      price: product.price.description,
      subscriptionPeriod: convertSKProductSubscriptionPeriod(toPigeon: product.subscriptionPeriod),
      introductoryPrice: convertProductDiscount(toPigeon: product.introductoryPrice),
      discounts: pigeonProductDiscounts)
  }

  public static func convertProductsResponse(toPigeon productsResponse: SKProductsResponse?)
    -> FIASKProductsResponseMessage?
  {
    guard let productsResponse = productsResponse else {
      return nil
    }

    let pigeonProducts = productsResponse.products.map { convertProduct(toPigeon: $0)! }

    return FIASKProductsResponseMessage.make(
      withProducts: pigeonProducts,
      invalidProductIdentifiers: productsResponse.invalidProductIdentifiers)
  }
}
